_posts/2023-05-26-correlated-subqueries-in-sql.md

```diff
@@ -0,0 +1,294 @@
+WHERE uniquecarrier = ANY (
+     SELECT uniquecarrier
+     FROM ontime
+     WHERE ontime.origin=ontime_outer.origin AND ontime.dest=ontime_outer.dest
+     GROUP BY uniquecarrier
+     HAVING COUNT(*) > 1000
+);
+```
+
+#### Performance
+
+Logically, correlated subqueries are executed *once per row*. As such, it is natural to think that correlated subqueries are very expensive and should be avoided for performance reasons.
+
+While that is true in many SQL systems, it is not the case in DuckDB. In DuckDB, subqueries are **always** *decorrelated*. DuckDB uses a state-of-the-art subquery decorrelation algorithm as described in the [Unnesting Arbitrary Queries](https://cs.emis.de/LNI/Proceedings/Proceedings241/383.pdf) paper. This allows all subqueries to be decorrelated and executed as a single, much more efficient, query.
+
+If we look at the query plan for the correlated scalar subquery using `EXPLAIN`, we can see that the query has been transformed into a hash aggregate followed by a hash join. This allows the query to be executed very efficiently.
+
+```sql
+EXPLAIN SELECT uniquecarrier, origincityname, destcityname, flightdate, distance
+FROM ontime AS ontime_outer
+WHERE distance=(
+     SELECT MIN(distance)
+     FROM ontime
+     WHERE uniquecarrier=ontime_outer.uniquecarrier
+);
+```
+
+```
+┌───────────────────────────┐
+│         HASH_JOIN         │ 
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │ 
+│      uniquecarrier =      │ 
+│       uniquecarrier       ├──────────────┐
+└─────────────┬─────────────┘              │
+┌─────────────┴─────────────┐┌─────────────┴─────────────┐
+│         SEQ_SCAN          ││       HASH_GROUP_BY       │
+│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ││   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+│           ontime          ││       uniquecarrier       │
+└───────────────────────────┘│       min(distance)       │
+                             └─────────────┬─────────────┘
+                             ┌─────────────┴─────────────┐
+                             │         SEQ_SCAN          │
+                             │   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
+                             │           ontime          │
+                             └───────────────────────────┘
+                             
+```
+
+We can see the drastic performance difference that subquery decorrelation has when we compare the run-time of this query in DuckDB with the run-time in Postgres and SQLite. When running the above query on the [`ontime` dataset](https://www.transtats.bts.gov/Homepage.asp) for `2017` with roughly `~4 million` rows, we get the following performance results:
+
+| DuckDB | Postgres | SQLite |
+|--------|----------|--------|
+| 0.06s  | >48 Hours    | >48 Hours  |
+
+
+As Postgres and SQLite do not de-correlate the subquery, the query is not just *logically*, but *actually* executed once for every row. As the subquery involves a full table scan, this moves the query from linear complexity, O(n), to quadratic complexity, O(n<sup>2</sup>).
```

Comment:
Maybe here we can add something along the lines of:
"Roughly speaking, this query will take 4 million times longer in Postgres/SQLite than in DuckDB."
I don't think the big O notation really resonates with everyone, but the number 4 million does.